### ISSUE TITLE
normalize_homography builds its normalization constants in float32: a float64 caller silently loses 8 significant digits (8.94e-09)

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#3958]
`normalize_homography`, `denormalize_homography` and `normalize_homography3d` build their normalization matrices by calling `normal_transform_pixel*` **without passing `dtype=` or `device=` through**, then casting the float32 result to the input:

```python
# kornia/geometry/conversions.py:1286 and :1291 (normalize_homography)
src_norm_trans_src_pix: torch.Tensor = normal_transform_pixel(src_h, src_w).to(dst_pix_trans_src_pix)
...
dst_norm_trans_dst_pix: torch.Tensor = normal_transform_pixel(dst_h, dst_w).to(dst_pix_trans_src_pix)
```

`normal_transform_pixel` defaults to float32, so the constants `2/(size − 1)` are rounded to float32 **before** the cast to float64. A float64 caller therefore gets roughly 8 significant digits of normalization constant, not 16 — the returned tensor has `dtype=torch.float64` while carrying float32-quality numbers.

## Repro

```python
import torch
from kornia.geometry.conversions import (
    denormalize_homography,
    normal_transform_pixel,
    normal_transform_pixel3d,
    normalize_homography,
    normalize_homography3d,
)

f64 = torch.float64
got = normalize_homography(torch.eye(3, dtype=f64)[None], (4, 4), (6, 6))[0]
Nd = normal_transform_pixel(6, 6, dtype=f64)[0]
Ns = normal_transform_pixel(4, 4, dtype=f64)[0]
exact = Nd @ torch.linalg.inv(Ns)

print("input dtype float64, output dtype:", got.dtype)
print("normalize_homography [0,0] =", got[0, 0].item())
print("float64-native       [0,0] =", exact[0, 0].item())
print("maxdiff                    =", (got - exact).abs().max().item())

print("\n# the leak at the source: the helper defaults to float32 whatever the caller uses")
print("normal_transform_pixel(4,5)            [1,1] =", normal_transform_pixel(4, 5)[0, 1, 1].item())
print("normal_transform_pixel(4,5,dtype=f64)  [1,1] =", normal_transform_pixel(4, 5, dtype=f64)[0, 1, 1].item())
print("exact 2/3                                    =", 2 / 3)

print("\n# same for the siblings")
g2 = denormalize_homography(torch.eye(3, dtype=f64)[None], (4, 4), (6, 6))[0]
e2 = torch.linalg.inv(Nd) @ Ns
print("denormalize_homography maxdiff =", (g2 - e2).abs().max().item())
g3 = normalize_homography3d(torch.eye(4, dtype=f64)[None], (2, 4, 5), (3, 8, 9))[0]
e3 = normal_transform_pixel3d(3, 8, 9, dtype=f64)[0] @ torch.linalg.inv(normal_transform_pixel3d(2, 4, 5, dtype=f64)[0])
print("normalize_homography3d maxdiff =", (g3 - e3).abs().max().item())
print("float64 eps for scale           =", torch.finfo(f64).eps)
```

## Observed

```
input dtype float64, output dtype: torch.float64
normalize_homography [0,0] = 0.5999999910593036
float64-native       [0,0] = 0.6000000000000001
maxdiff                    = 8.94069651646845e-09

# the leak at the source: the helper defaults to float32 whatever the caller uses
normal_transform_pixel(4,5)            [1,1] = 0.6666666865348816
normal_transform_pixel(4,5,dtype=f64)  [1,1] = 0.6666666666666666
exact 2/3                                    = 0.6666666666666666

# same for the siblings
denormalize_homography maxdiff = 2.483526828633842e-08
normalize_homography3d maxdiff = 6.386211781617135e-09
float64 eps for scale           = 2.220446049250313e-16
```

The deviation is `~9e-09` against a float64 epsilon of `2.2e-16` — seven orders of magnitude of avoidable error, present on every float64 call.

## Expected

`normalize_homography(H, ...)` with a float64 `H` should produce float64-quality constants: `[0,0] = 0.6000000000000001`, not `0.5999999910593036`. The helper already accepts `dtype=` and `device=`, so the fix is to pass them through:

```python
src_norm_trans_src_pix = normal_transform_pixel(
    src_h, src_w, device=dst_pix_trans_src_pix.device, dtype=dst_pix_trans_src_pix.dtype
)
```

Why it is worth fixing rather than shrugging at:

- It defeats the *point* of running in float64. A user who deliberately picks double precision for a geometry pipeline — bundle adjustment, calibration, anything where homographies get composed or inverted repeatedly — receives silently degraded constants and has no way to see it from the returned dtype.
- The error enters at the *first* step of a chain and is then amplified by every subsequent composition and inverse.
- The helper matrices are also materialized on CPU by default and cast to the input, so the same missing pass-through means a device round-trip on every non-CPU call. (I did not measure the copy — no profiler trace was run — so I state it as a code consequence, not a measured cost. The dtype half above *is* measured.)

**Scope note:** this touches `normalize_homography`'s helper call site, which is in the same function as **#3904**'s territory. It does **not** touch the `2/(size − 1)` scaling formula or the `align_corners` convention — only the precision and placement at which the existing formula is materialized — so the two changes do not conflict, but a PR for this should be scoped narrowly and say so.

## Classification

**Broken output with a default-changing fix.** Returning float32-rounded constants in a float64 tensor is wrong under any reading, and the fix is additive in quality. But it changes the returned values of every float64 call by `~1e-08`, so it cannot land under byte-identity discipline for float64 (float32 and float16 paths *are* byte-identical before and after, since the cast is then a no-op or a widening of already-float32 constants — worth verifying in the PR).

Per the reference-core lifecycle (Part I / Part III §2) my reading is that a `1e-08 → 1e-16` improvement in float64 has no plausible downstream dependent and can land as an ordinary bugfix, but since it is technically a default change I am flagging the choice rather than deciding it.

Related closed prior art, cited but **not** claimed as the same cause: **#1769** and **#888** both report `normalize_homography` producing values that deviate from the expected identity (CUDA-specific, torch 1.10 era). I have not reproduced either — no CUDA is available on this host — and their magnitudes are much larger than the float32-rounding effect measured above, so they are probably a different problem. Mentioned only so a triager does not spend time rediscovering them.

## Environment

kornia 0.9.0rc1 @ `3358ef16`, torch 2.9.1, cpu (no CUDA on this host — no CUDA claim is made). Executed from a worktree carrying batch-3a docs/tests commits on top of `3358ef16`; `normalize_homography`, `denormalize_homography`, `normalize_homography3d` and both `normal_transform_pixel*` helpers are byte-identical to `origin/main` there (no diff hunk lands in any of them; function bodies compare AST-identical).

Posted on behalf of @ducha-aiki by Claude (Fable 5).

**Comment by aswanth-07** [comment]
Hi, I'd like to work on this. I plan to keep the fix scoped to constructing the existing homography normalization transforms directly with the input tensor's dtype and device, then add focused float64 regression tests for the affected 2D and 3D paths while confirming float32 behavior remains unchanged. Could a maintainer please confirm the scope and assign the issue to me before I start, per the contribution policy?

**Comment by ducha-aiki** [comment]
@aswanth-07 go ahead, please. Keep in mind that those “fix” PRs will be merged only after current audit phase.

**Comment by aswanth-07** [comment]
Thanks for confirming. I found that #3979 already implements this same dtype/device fix, so I'll step back rather than duplicate the existing work.
